irohad/subscription/dispatcher.hpp

```diff
@@ -12,90 +12,27 @@
 namespace iroha::subscription {
 
   template <uint32_t kCount, uint32_t kPoolSize>
-  class Dispatcher final : utils::NoCopy, utils::NoMove {
-   public:
-    static constexpr uint32_t kHandlersCount = kCount;
-    static constexpr uint32_t kPoolThreadsCount = kPoolSize;
-    using Task = ThreadHandler::Task;
+  struct IDispatcher {
     using Tid = uint32_t;
-    static constexpr Tid kExecuteInPool = std::numeric_limits<Tid>::max();
-
-   private:
-    struct ThreadHandlerContext {
-      /// Worker thread to execute tasks
-      std::shared_ptr<ThreadHandler> handler;
-
-      /// Shows if this handler is static or if it was created to
-      /// execute a single task and should be deleted after performing it
-      bool is_temporary;
-    };
-
-    ThreadHandlerContext handlers_[kHandlersCount];
-    ThreadHandlerContext pool_[kPoolThreadsCount];
-
-    inline ThreadHandlerContext findHandler(Tid const tid) {
-      assert(tid < kHandlersCount || tid == kExecuteInPool);
-      if (tid < kHandlersCount)
-        return handlers_[tid];
-
-      for (auto &handler : pool_)
-        if (!handler.handler->isBusy())
-          return handler;
-
-      return ThreadHandlerContext{
-          std::make_shared<ThreadHandler>(),
-          true  // temporary
-      };
-    }
+    using Task = ThreadHandler::Task;
 
-   public:
-    Dispatcher() {
-      for (auto &h : handlers_) {
-        h.handler = std::make_shared<ThreadHandler>();
-        h.is_temporary = false;
-      }
-      for (auto &h : pool_) {
-        h.handler = std::make_shared<ThreadHandler>();
-        h.is_temporary = false;
-      }
-    }
+    static constexpr Tid kExecuteInPool = std::numeric_limits<Tid>::max();
+    static constexpr uint32_t kHandlersCount = kCount;
```

Comment:
IDispatcher should not have these parameters. They are not applicable with MockDispatcher, or single threaded dispatcher, for example.